db/submission_test.go

```diff
@@ -59,3 +62,72 @@ func (this *DBTests) DBTestRemoveSubmission(test *testing.T) {
         }
     }
 }
+
+func (this *DBTests) DBTestFetchAttempts(test *testing.T) {
+    ResetForTesting();
+    defer ResetForTesting();
+
+    assignment := MustGetTestAssignment();
+
+    // Fetch attempts from another user who has submissions.
+    studentAttempts, err := GetSubmissionAttempts(assignment, "student@test.com");
+    if err != nil {
+        test.Fatalf("Failed to get student attempts when should be a success: %v", err);
+    }
+
+    if len(studentAttempts) == 0 {
+        test.Fatalf("Got an empty result when shouldn't be.");
+    }
+
+    // Fetch attempts from self (grader). User does not have submissions.
+    graderAttempts, err := GetSubmissionAttempts(assignment, "grader@test.com");
+    if err != nil {
+        test.Fatalf("Failed to get grader attempts when should be a success (with empty result): %v", err);
+    }
+
+    if len(graderAttempts) != 0 {
+        test.Fatalf("Got a non-empty result when should be empty.");
+    }
+
+    // User makes a submission, then that submission is removed.
+    // This case tests GetSubmissionAttempts() when the submission directory exists but is empty.
```

Comment:
The comment on this line (93) is a bit misplaced.
That is the general idea of this test function, but this tries to be specific to a test case.
But then this is the wrong test case for it.

So instead of this line, put a comment on the function (right above the `func` line) that lays out the general flow of this test function.
e.g., get submissions from another user, check for zero, hijack a submission, check for 1, remove, check for zero.
I also wouldn't bother mentioning the directory, since that is specific to the disk db.

This will probably also change some of your other comments in this function.